README.md
# SSScheduler
Smart Storage Scheduler for 2022 Massive Storage Competition

## train of thoughts v1.0:

![](./images/algorithm.png)

## alns:
<font color=#ff0000>score: 399.5 -> 490.0</font>
![](./images/alns.png)

- BPR算子：随机选择一个driver，选择其中一个request（根据它的requests的size来设置概率，size越大的被选择概率越大），将该request移除，从等待队列里面重新选择request放入该driver中。

- shift算子：随机选择一个driver，根据上述概率选择一个request移到另一个driver上面去，如果目标driver的capacity超出，则先选择一部分他的request移到其它driver上面去，如果依然超出，则移除部分requests，知道capacity不超出为止，最后从等待队列里面重新选择request放入这两个driver中。

- 退火算法计算接收概率：
  ![](./images/anneal.png)

- 一定要防止提交的driver的capacity被超出！为此我修改了runner.py，当出现capacity超出的情况会直接报错停止运行。


history/history.txt
-609.0
-541.0
-476.0
399.5
475.5
486.5


alns.py
from typing import List
import random
import copy
from time import perf_counter
import string
from math import exp, floor

class ReqStructure():
    def __init__(self) -> None:
        self.RequestID: int
        self.RequestType: string
        self.SLA: int
        self.Driver: List[int]
        self.RequestSize: int
        self.LogicalClock: int
        # parameters define by ourselves
        self.now_sla: int
        self.type: string
        self.score: float
        self.selected_driver: int

# 尚不确定是否需要state
class state():
    def __init__(self) -> None:
        # req_state format: 
        # -1 means can't put in driver, other number means driver's id, 
        # e.g. [-1, -1, 0, 2, -1] means req[2] and req[3] are put into driver 0 and driver 2, 
        # others can't be put into any drivers.
        self.req_state: List[int] = []
        self.solu_score: int = 0

class OperReq():
    """
    a structure which is convenient for operation.
    """
    def __init__(self) -> None:
        self.Driver: List[int]
        self.RequestSize: int
        self.score: float
        self.id: int
        self.selected_driver: int

class alns():
    def __init__(self, ini_requests:List[ReqStructure], remain_cap:List[int], 
    ini_score:float, max_iteraion:int, max_runtime:int, real_cap:List[int], 
    start_temp:float, end_temp:float, temp_step:float, temp_s1:float) -> None:
        """
        ini_requests: requests dealed with wfac-bin-packing
        remain_cap: driver remain capacity after wfac-bin-packing
        ini_score: sum of score about wfac-bin-packing
        real_cap: real max capacity of a driver
        """
        self.num_oper = 2
        self.num_driver = len(remain_cap)
        self.real_cap = copy.deepcopy(real_cap)
        self.max_iteration = max_iteraion
        self.max_runtime = max_runtime
        self.oper_weight: List[int] = [1] * self.num_oper
        self.oper_prob: List[float] = [1 / self.num_oper] * self.num_oper
        self.iteraion = 0
        self.runtime = 0

        self.start_temp = start_temp
        self.end_temp = end_temp
        self.temp = copy.deepcopy(self.start_temp)
        self.step = temp_step
        self.s1 = temp_s1    # ajust to bigger, acception prob will be smaller

        self.best_score = copy.deepcopy(ini_score)    # score of best solution
        self.score = copy.deepcopy(ini_score)         # score of previous solution
        self.best_reqs = copy.deepcopy(ini_requests)
        self.requests = copy.deepcopy(ini_requests)    # reqs of previous solution, can't change its sequence, only change its selected_driver
        self.best_remain_cap = copy.deepcopy(remain_cap)
        self.remain_cap = copy.deepcopy(remain_cap)
        self.best_state = copy.deepcopy(self.reqs2state(ini_requests))
        self.state = copy.deepcopy(self.best_state)    # state of previous solution

        self.up_reqs: List[List[OperReq]] = []   # reqs on each driver, only use it in operations
        self.down_reqs: List[OperReq] = [] # reqs waiting, only use it in operations, sort from big to small according to their score
        self.ini_on_down_req()
    
    def iteration_alns(self):
        start_time = perf_counter()
        while not self.is_stop():
            oper_id = self.choose_prob(self.oper_prob) 
            cand_up_reqs, cand_down_reqs, cand_remain_cap, cand_score = self.operaion(oper_id=oper_id)
            if self.is_accepted(score=cand_score):
                # update parameters
                self.up_reqs = cand_up_reqs
                self.down_reqs = cand_down_reqs
                self.remain_cap = cand_remain_cap
                self.score = cand_score
                self.OperReq2ReqStrut(cand_up_reqs=cand_up_reqs, cand_down_reqs=cand_down_reqs) # update self.requests
                self.update_weight_prob(oper_id=oper_id, step=1)
            if cand_score > self.best_score:
                self.best_reqs = copy.deepcopy(self.requests)
                self.best_score = copy.deepcopy(self.score)
                self.best_remain_cap = copy.deepcopy(self.remain_cap)
                self.update_weight_prob(oper_id=oper_id, step=5)

            self.iteraion += 1
            self.runtime = perf_counter() - start_time
            self.temp = max(self.end_temp, self.temp * self.step)
        return self.best_reqs, self.best_score, self.best_remain_cap
            
    
    def update_weight_prob(self, oper_id:int, step:int):
        self.oper_weight[oper_id] += step
        s = sum(self.oper_weight)
        self.oper_prob = [float(w / s) for w in self.oper_weight]

    def choose_prob(self, prob:List[float]) -> int:
        """
        choose a element according to probability list.

        return: element index
        """
        p = random.random()
        s = 0
        for i in range(len(prob)):
            s += prob[i]
            if s >= p:
                return i
        return 0

    def ini_on_down_req(self):
        for _ in range(len(self.remain_cap)):
            self.up_reqs.append([])
        for i in range(len(self.requests)):
            r = self.requests[i]
            oper_req = OperReq()
            oper_req.id = i
            oper_req.Driver = r.Driver
            oper_req.RequestSize = r.RequestSize
            oper_req.score = r.score
            oper_req.selected_driver = r.selected_driver
            if r.selected_driver != -1:
                self.up_reqs[r.selected_driver].append(oper_req)
            else:
                self.down_reqs.append(oper_req)

    def OperReq2ReqStrut(self, cand_up_reqs:List[List[OperReq]], cand_down_reqs:List[OperReq]):
        """
        change the self.requests' selected_driver  through cand_up_reqs and cand_down_reqs
        """
        for driver_reqs in cand_up_reqs:
            for r in driver_reqs:
                self.requests[r.id].selected_driver = r.selected_driver
        for r in cand_down_reqs:
            self.requests[r.id].selected_driver = r.selected_driver

    def operaion(self, oper_id: int):
        """
        excuting opraion through the oper_id.
        """
        if oper_id == 0:
            return self.operation_0()
        elif oper_id == 1:
            return self.operation_1()
        else:
            raise IndexError("Error: operaion id is not exist")

    def reqs2state(self, reqs:List[ReqStructure]) -> state:
        """
        get the reqs' state

        return: state 
        """
        s = state()
        for r in reqs:
            s.req_state.append(r.selected_driver)
            s.solu_score += r.score
        return s

    def is_accepted(self, score:float) -> bool:
        """
        judging whether the candidate solution is need to be accepted,
        we use hill climbing in there.

        return: true means the solution is accepted.
        """
        if score >= self.score:
            return True
        else:
            d_score = score - self.score
            p = exp((d_score / self.temp) * self.s1)
            return random.random() <= p 

    def is_stop(self):
        """
        judging whether the iteraion solution is need to be stopped.
        """
        if self.max_iteration > self.iteraion and self.max_runtime > self.runtime:
            return False
        else:
            return True

    
    def prob_ger_req(self, req_id_list:List[int], driver_reqs:List[OperReq]) -> int:
        """
        compute each req's probability be chosen in the list, according to their RequestSize, 
        and then get a req through the probability list.

        return: req_id in reqs list
        """
        try:
            sum_size = sum([driver_reqs[i].RequestSize for i in req_id_list])
            prob = [float(driver_reqs[i].RequestSize / sum_size) for i in req_id_list]
            return req_id_list[self.choose_prob(prob=prob)]
        except:
            return req_id_list[0]

    def move_to_other(self, req:OperReq, driver_id:int) -> bool:
        """
        judge if a req's size is bigger than other driver's max_capacity except current driver, 
        if it is, it can't move to other driver.

        return: true means it can move 
        """
        for d in req.Driver:
            if d != driver_id and self.real_cap[d] > req.RequestSize:
                return True
        return False

    def up_add_req(self, driver_cand_up_reqs:List[OperReq], req:OperReq, driver_id:int, driver_cap:int, score:float):
        """
        add new req to a driver's up_reqs list, and update parameters

        return: driver_cand_up_reqs: List[OperReq], driver_cap: int, score: float
        """
        req.selected_driver = driver_id
        driver_cand_up_reqs.append(req)
        driver_cap -= req.RequestSize
        score += req.score
        # if driver_cap < 0:
        #     raise AttributeError(f'up_add_req: {driver_cap}')
        return driver_cand_up_reqs, driver_cap, score

    def up_remove_req(self, driver_cand_up_reqs:List[OperReq], req:OperReq, driver_cap:int, score:float):
        """
        remove a req from a driver's up_reqs list, and update parameters

        return: driver_cand_up_reqs: List[OperReq], driver_cap: int, score: float
        """
        driver_cand_up_reqs.remove(req)
        driver_cap += req.RequestSize
        score -= req.score
        # if driver_cap < 0:
        #     raise AttributeError(f'up_remove_req: {driver_cap}')
        return driver_cand_up_reqs, driver_cap, score

    def down_add_req(self, cand_down_reqs:List[OperReq], req:OperReq):
        """
        add a new req to cand_down_reqs

        return: cand_down_reqs:List[OperReq] 
        """
        s = []
        for i in range(len(cand_down_reqs)):
            if req.score > cand_down_reqs[i].score:
                if i == 0:
                    req.selected_driver = -1; s.append(req); cand_down_reqs = s + cand_down_reqs; break
                else:
                    req.selected_driver = -1; s = cand_down_reqs[:i]; s.append(req); cand_down_reqs = s + cand_down_reqs[i:]; break
        if req.selected_driver != -1:
            req.selected_driver = -1; cand_down_reqs.append(req)
        return cand_down_reqs

    def down_remove_req(self, cand_down_reqs:List[OperReq], driver_cand_up_reqs:List[OperReq], driver_id:int, driver_cap:int, cand_score:int):
        """
        move some reqs of waiting list to a driver's up_reqs lists

        return: cand_down_reqs: List[OperReq], driver_cand_up_reqs: List[OperReq], driver_cap: int, cand_score: int
        """
        for i in range(len(cand_down_reqs)):
            if driver_id in cand_down_reqs[i].Driver and cand_down_reqs[i].RequestSize < driver_cap:
                cand_down_reqs[i].selected_driver = driver_id
                driver_cand_up_reqs.append(cand_down_reqs[i])
                driver_cap -= cand_down_reqs[i].RequestSize
                cand_score += cand_down_reqs[i].score
        # delete reqs move to drivers
        for i in range(len(cand_down_reqs)-1, -1, -1):
            if cand_down_reqs[i].selected_driver != -1:
                cand_down_reqs.remove(cand_down_reqs[i])
        # if driver_cap < 0:
        #     raise AttributeError(f'down_remove_req: {driver_cap}')
        return cand_down_reqs, driver_cand_up_reqs, driver_cap, cand_score

    def operation_0(self):
        """
        BPR(Big Process Remove) operation:
        for one driver, put a big size req down, try to put waiting reqs to this driver

        return: candidate driver_reqs: List[List[req]], candidate down_reqs: List[req], 
                candidate driver_remain_cap: List[int], candidate score: float
        """
        # create candidate solution 
        cand_up_reqs = copy.deepcopy(self.up_reqs)
        cand_down_reqs = copy.deepcopy(self.down_reqs)
        cand_remain_cap = copy.deepcopy(self.remain_cap)
        cand_score = copy.deepcopy(self.score)

        driver_id = random.randint(0, self.num_driver-1)

        if len(cand_up_reqs[driver_id]) > 1:
            num_deal_req = max(1, random.choice([1, floor(0.25 * len(cand_up_reqs[driver_id])), \
                floor(0.5 * len(cand_up_reqs[driver_id]))]))
        elif len(cand_up_reqs[driver_id]) == 1:
            num_deal_req = 1
        else:
            return [], [], [], -100 # if the driver has not any reqs match condition. return immediately
        
        req_id_list: List[int] = list(range(len(cand_up_reqs[driver_id])))
        req_list: List[OperReq] = []
        for _ in range(num_deal_req):
            req_id = self.prob_ger_req(req_id_list, cand_up_reqs[driver_id])
            req = cand_up_reqs[driver_id][req_id]
            req_list.append(req)
            req_id_list.remove(req_id)
        
        for req in req_list:
            cand_up_reqs[driver_id], cand_remain_cap[driver_id], cand_score = \
                self.up_remove_req(cand_up_reqs[driver_id], req, cand_remain_cap[driver_id], cand_score)
        # down_reqs move req to driver
        cand_down_reqs, cand_up_reqs[driver_id], cand_remain_cap[driver_id], cand_score = \
            self.down_remove_req(cand_down_reqs, cand_up_reqs[driver_id], driver_id, cand_remain_cap[driver_id], cand_score)
        # insert req into down_reqs in order
        for req in req_list:
            cand_down_reqs = self.down_add_req(cand_down_reqs, req)

        for c in cand_remain_cap:
            if c < 0:
                raise AttributeError(f'oper_0: {cand_remain_cap}')

        return cand_up_reqs, cand_down_reqs, cand_remain_cap, cand_score
        

    def operation_1(self) -> List[ReqStructure]:
        """
        shift operation:
        for one driver, put a big size req move to another driver, 
        if the target driver capacity is over, try to move some reqs to other drivers, or put into waiting reqs list,
        after that, try to put waiting reqs to two drivers.

        return: candidate driver_reqs: List[List[req]], candidate down_reqs: List[req], 
                candidate driver_remain_cap: List[int], candidate score: float
        """
        # create candidate solution 
        cand_up_reqs = copy.deepcopy(self.up_reqs)
        cand_down_reqs = copy.deepcopy(self.down_reqs)
        cand_remain_cap = copy.deepcopy(self.remain_cap)
        cand_score = copy.deepcopy(self.score)

        driver_id = random.randint(0, self.num_driver-1)
        # find a req in driver's reqs according to its size, and remove it
        max_req_list: List[OperReq] = []
        for i in range(len(cand_up_reqs[driver_id])):
            r = cand_up_reqs[driver_id][i]
            if len(r.Driver) > 1 and self.move_to_other(r, driver_id):
                max_req_list.append(r)
        if max_req_list == []:  # if the driver has not any reqs match condition. return immediately
            return [], [], [], -100

        if len(max_req_list) > 1:
            num_deal_req = max(1, random.choice([1, floor(0.25 * len(max_req_list)), floor(0.5 * len(max_req_list))]))
        else:
            num_deal_req = 1
        
        req_id_list: List[int] = list(range(len(max_req_list)))
        req_list: List[OperReq] = []
        for _ in range(num_deal_req):
            req_id = self.prob_ger_req(req_id_list, max_req_list)
            req = max_req_list[req_id]
            req_list.append(req)
            req_id_list.remove(req_id)

        removed_r: List[OperReq] = [] # collet reqs be removed by to_driver_id
        # move chosen reqs from driver_id to to_driver_ids 
        for max_req in req_list:
            # driver id remove max_req
            cand_up_reqs[driver_id], cand_remain_cap[driver_id], cand_score = \
                self.up_remove_req(cand_up_reqs[driver_id], max_req, cand_remain_cap[driver_id], cand_score)

            # find the driver max_req moved to
            max_req_driverlist = copy.deepcopy(max_req.Driver)
            max_req_driverlist.remove(driver_id)
            if len(max_req_driverlist) > 1:
                for i in range(len(max_req_driverlist)-1, -1, -1):
                    d = max_req_driverlist[i]
                    if self.real_cap[d] < max_req.RequestSize:
                        max_req_driverlist.remove(d)
                to_driver_id = random.choice(max_req_driverlist)
            else:
                to_driver_id = max_req_driverlist[0]
            if self.real_cap[to_driver_id] < max_req.RequestSize:
                raise AttributeError('self.real_cap[to_driver_id] < max_req.RequestSize')
            
            cand_up_reqs[to_driver_id], cand_remain_cap[to_driver_id], cand_score = \
                self.up_add_req(cand_up_reqs[to_driver_id], max_req, to_driver_id, cand_remain_cap[to_driver_id], cand_score)
            
            # if to_driver_id's capacity is too less, try to move some reqs to other drivers
            if cand_remain_cap[to_driver_id] < 0:
                for i in range(len(cand_up_reqs[to_driver_id])-1, -1, -1):
                    if cand_remain_cap[to_driver_id] >= 0:
                        break

                    r = cand_up_reqs[to_driver_id][i]
                    if r.id == max_req.id:
                        continue

                    if len(r.Driver) > 1:
                        max_cap = -100
                        max_d = -1
                        for d in r.Driver:
                            if cand_remain_cap[d] > r.RequestSize and d != to_driver_id and cand_remain_cap[d] > max_cap:
                                max_cap = cand_remain_cap[d]
                                max_d = d
                        if max_d != -1: 
                            # move the request r from to_driver_id to max_d
                            cand_up_reqs[to_driver_id], cand_remain_cap[to_driver_id], cand_score = \
                                self.up_remove_req(cand_up_reqs[to_driver_id], r, cand_remain_cap[to_driver_id], cand_score)
                            cand_up_reqs[max_d], cand_remain_cap[max_d], cand_score = \
                                self.up_add_req(cand_up_reqs[max_d], r, max_d, cand_remain_cap[max_d], cand_score)
            # if to_driver_id's capacity is still too less, remove some reqs to the down list
            if cand_remain_cap[to_driver_id] < 0:
                for i in range(len(cand_up_reqs[to_driver_id])-1, -1, -1):
                    r = cand_up_reqs[to_driver_id][i]
                    if cand_remain_cap[to_driver_id] >= 0: break
                    if r.id == max_req.id: continue
                    cand_up_reqs[to_driver_id], cand_remain_cap[to_driver_id], cand_score = \
                        self.up_remove_req(cand_up_reqs[to_driver_id], r, cand_remain_cap[to_driver_id], cand_score)
                    removed_r.append(r)
            if cand_remain_cap[to_driver_id] < 0:
                raise AttributeError(f'{cand_remain_cap}, num_req:{len(cand_up_reqs[to_driver_id])}')

        # move reqs from down to up
        for i in range(len(cand_down_reqs)):
            max_cap = -100
            max_d = -1
            for d in [driver_id, to_driver_id]:
                if d in cand_down_reqs[i].Driver and cand_remain_cap[d] > cand_down_reqs[i].RequestSize and \
                    cand_remain_cap[d] > max_cap:
                    max_cap = cand_remain_cap[d]
                    max_d = d
            if max_d != -1:
                cand_down_reqs[i].selected_driver = max_d
                cand_up_reqs[max_d], cand_remain_cap[max_d], cand_score = \
                    self.up_add_req(cand_up_reqs[max_d], cand_down_reqs[i], max_d, cand_remain_cap[max_d], cand_score)
        for i in range(len(cand_down_reqs)-1, -1, -1):
            if cand_down_reqs[i].selected_driver != -1:
                cand_down_reqs.remove(cand_down_reqs[i])
        
        # reqs be remove by to_driver_id, insert into down reqs
        for r in removed_r:
                cand_down_reqs = self.down_add_req(cand_down_reqs, r)
        
        for c in cand_remain_cap:
            if c < 0:
                raise AttributeError(f'oper_2: {cand_remain_cap}')

        return cand_up_reqs, cand_down_reqs, cand_remain_cap, cand_score





scheduler.py
import os
import sys
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__)) # __file__如果不行，就改成'file'
sys.path.append(os.path.dirname(SCRIPT_DIR))
import string
from abc import ABCMeta, abstractmethod
import json
from typing import List
import math
from memory_profiler import psutil
import copy
from alns import alns
from concurrent import futures

# worst score: -172600.5
# now score: -609.0
# now score: -541.0
# now score: -476.0
# now score: 399.5
# now score: 475.5
# now score: 485.5, 496.0, 487.5
# now score: 548.0

# num_thread: 1  self.incre_score: 128.0  score: 497.5
# num_thread: 4  self.incre_score: 160.0  score: 476.0

URGENT = "URGENT"
noURGENT = "noURGENT"
ABANDON = "ABANDON"

class Scheduler(metaclass=ABCMeta):
    @abstractmethod
    def init(self, driver_num: int) -> None:
        pass

    @abstractmethod
    def schedule(self, logical_clock: int, request_list: list, driver_statues: list) -> list:
        pass

class ReqStructure():
    def __init__(self) -> None:
        self.RequestID: int
        self.RequestType: string
        self.SLA: int
        self.Driver: List[int]
        self.RequestSize: int
        self.LogicalClock: int
        # parameters define by ourselves
        self.now_sla: int
        self.type: string
        self.score: float
        self.selected_driver: int

class DriverStructure():
    def __init__(self) -> None:
        self.DriverID: int
        self.Capacity: int
        self.LogicalClock: int


class DemoScheduler(Scheduler):
    def __init__(self):
        self.requests: List[ReqStructure] = []
        self.driver_status: List[DriverStructure] = []
        self.remain_cap: List[int] = []
        self.real_cap: List[int] = []
        self.memory: List[int] = []
        self.driver_num = 0
        self.num_URGENT = 0
        self.logical_clock = 0
        self.score = 0
        self.incre_score = 0

        # parameters
        self.max_iteration = 1000 
        self.max_runtime = 1.5
        self.start_temp = 1500 # 100
        self.end_temp = 10  # 10
        self.temp_step = 0.99 # 0.997 0.99
        self.temp_s1 = 180     # 10 # ajust to bigger, acception prob will be smaller
        self.num_thread = 4     # alns thread
        self.nourgent_addition = 3 # if nourgent's sla <= self.nourgent_addition, it will get an addtion parameter
        self.para_alpha_add = 5  # nourgent's addition parameter

    def init(self, driver_num: int) -> None:
        self.driver_num = driver_num

    def get_memory(self) -> int:
        """
        get memory process has occupied
        """
        return psutil.Process(os.getpid()).memory_info().rss / 1024 / 1024
        
    def set_type(self, request:ReqStructure) -> string:
        """
        set request's type to URGENT or noURGENT
        return: type
        """
        if request.RequestSize == 0:
            return ABANDON
        if request.RequestType == "BE" and request.now_sla < 1: # if BE overtime, we can't get score anymore
            return ABANDON

        if request.now_sla > 1 :
            return noURGENT
        elif request.now_sla <= 1 and request.now_sla >= -10:  
            return URGENT 
        else:   
            # "FE" or "EM" <=-11 means req has over 12 hours, 
            # req has get biggest deduction, we can abandon it immediately
            return ABANDON

    def set_score(self, request:ReqStructure) -> int:
        """
        computing score of the request

        return: score
        """
        if request.type == URGENT:
            if request.now_sla > 1: # set type error
               return 0
            # minus_hours = -(request.now_sla-2)
            if request.RequestType == "FE":
                return math.ceil(request.RequestSize / 50)
            elif request.RequestType == "BE":
                return 0.5 * math.ceil(request.RequestSize / 50)
            else:
                return 2 *  math.ceil(request.RequestSize / 50)
       
        else:
            if request.now_sla <= 1: # set type error
                return 10000
            elif request.now_sla > 1 and request.now_sla <= 3: # sla = 2 or 3, its socre will be doubled
                alpha_addition = 8
            elif request.now_sla > 3 and request.now_sla <= 6:
                alpha_addition = 2
            else:
                alpha_addition = 1
                
            if request.RequestType == "FE":
                return  math.ceil(request.RequestSize / 50) * alpha_addition
            elif request.RequestType == "BE":
                return  0.5 * math.ceil(request.RequestSize / 50) * alpha_addition
            else:
                return  2 * math.ceil(request.RequestSize / 50) * alpha_addition


    def sort(self):
        """
        computing request's score and insert it to request dict, 
        and then sort requests.
        """
        # divide requests list into two lists by type
        num_urg = 0
        urg_requests: List[ReqStructure] = []
        nourg_requests: List[ReqStructure] = []
        for r in self.requests:
            if r.type == URGENT:
                urg_requests.append(r)
                num_urg += 1
            else:
                nourg_requests.append(r)
        self.num_URGENT = num_urg

        # sort URGENT firstly, high score, small size requests are in the frontier
        for i in range(len(urg_requests)):
            for j in range(i+1, len(urg_requests)):
                if urg_requests[i].score < urg_requests[j].score or \
                  (urg_requests[i].score == urg_requests[j].score and \
                    urg_requests[i].RequestSize > urg_requests[j].RequestSize):
                    tmp = urg_requests[i]
                    urg_requests[i] = urg_requests[j]
                    urg_requests[j] = tmp
        # sort noURGENT secondly
        for i in range(len(nourg_requests)):
            for j in range(i+1, len(nourg_requests)):
                if nourg_requests[i].score < nourg_requests[j].score or \
                  (nourg_requests[i].score == nourg_requests[j].score and \
                    nourg_requests[i].RequestSize > nourg_requests[j].RequestSize):
                    tmp = nourg_requests[i]
                    nourg_requests[i] = nourg_requests[j]
                    nourg_requests[j] = tmp
        
        self.requests = urg_requests + nourg_requests
        # print(f'after sort, self.request:')
        # for r in self.requests:
        #     print(r)

    def select_type(self, type:string) -> list:
        """
        get one type of requests from the self.requests

        return: list of reqs belong to the type
        """
        if type == URGENT:
            return self.requests[:self.num_URGENT]
        else:
            return self.requests[self.num_URGENT:]

    def set_type_reqs(self, results:list, type:string):
        """
        after the results is certain, 
        set one type of requests' selected_driver from the self.requests
        """
        reqs_len = len(self.requests)
        if type == URGENT:
            self.requests[:self.num_URGENT] = copy.deepcopy(results)
        else:
            self.requests[self.num_URGENT:] = copy.deepcopy(results)
        assert len(self.requests) == reqs_len

    def wfac_algo(self, requests:List[ReqStructure], driver_cap:list):
        """
        wfac bin-packing algorithm

        return: results(set the selected_driver of all self.requests), 
                sum of score, remain capacity of drivers
        """
        assert len(driver_cap) == self.driver_num

        score = 0
        results = copy.deepcopy(requests)
        driver_remain = copy.deepcopy(driver_cap)
        for i in range(len(results)):
            # find if there are drivers can hold the request, 
            # if true, select the driver with biggest capcity,
            # set the request's "Driver" as [driver_id].
            max_cap = 0
            driver = -1
            for d in results[i].Driver:
                if driver_remain[d] > results[i].RequestSize and driver_remain[d] > max_cap:
                    max_cap = driver_remain[d]
                    driver = d
            if driver != -1:
                results[i].selected_driver = driver
                driver_remain[driver] = driver_remain[driver] - results[i].RequestSize
                score += results[i].score
        # print('results:')
        # for r in results:
        #     print(r.selected_driver)
        return results, score, driver_remain

    def alns_algo(self, wfac_results:List[ReqStructure], wfac_remain_cap:List[int], wfac_score:int, real_cap:List[int]):
        alns_al = alns(ini_requests=wfac_results, remain_cap=wfac_remain_cap, ini_score=wfac_score, \
                max_iteraion=self.max_iteration, max_runtime=self.max_runtime, real_cap=real_cap, \
                start_temp=self.start_temp, end_temp=self.end_temp, temp_step=self.temp_step, temp_s1=self.temp_s1)
        return alns_al.iteration_alns()

    def type_schedule(self, type:string):
        """
        make schedule for a type of requests, 
        don't forget delete requests will be commited in self.requests

        return: some requests in self.requests's selected_driver has been sure 
        """
        # run all algoritym and get the best results
        assert len(self.remain_cap) == self.driver_num

        requests = self.select_type(type)
        wfac_results, wfac_score, wfac_remain_cap = self.wfac_algo(requests, self.remain_cap)

        if type == URGENT:
            real_cap = self.real_cap
        else:
            real_cap = self.remain_cap
        # try:
        num_thread = self.num_thread
        pool = futures.ThreadPoolExecutor(max_workers=num_thread)
        task: List[futures._base.Future] = []
        for i in range(num_thread):
            task.append(pool.submit(self.alns_algo, wfac_results, wfac_remain_cap, wfac_score, real_cap))

        alns_results = []; alns_score = -100; alns_remain_cap = []
        done_list = []
        num_done_thread = 0
        while True:
            for i in range(num_thread):
                if i not in done_list:
                    if task[i].done():
                        num_done_thread += 1
                        done_list.append(i)
                        results, score, remain_cap = task[i].result()
                        print(score)
                        if score >  alns_score:
                            alns_results = copy.deepcopy(results)
                            alns_score = score
                            alns_remain_cap = copy.deepcopy(remain_cap)
            if num_done_thread == num_thread:
                break

        pool.shutdown()
        # raise AttributeError(f'only wfac')
        # except:
        #     self.remain_cap = wfac_remain_cap
        #     self.set_type_reqs(wfac_results, type)
        #     self.score = wfac_score
        #     return
        
        self.memory.append(self.get_memory())
        print(f'********wfac_score:{wfac_score}********')
        print(f'********alns_score:{alns_score}********')
        # alns_score = -1000
        if wfac_score >= alns_score:
            self.remain_cap = copy.deepcopy(wfac_remain_cap)
            self.set_type_reqs(wfac_results, type)
            self.score = wfac_score
        else:
            self.remain_cap = copy.deepcopy(alns_remain_cap)
            self.set_type_reqs(alns_results, type)
            self.score = alns_score
            self.incre_score += (alns_score - wfac_score)
        print(f'self.incre_score: {self.incre_score}')

    def excu_reqs(self) -> list:
        """
        translate requests will be commited into formal excu_reqs of drivers

        return: excu_reqs
        """
        all_driver_req = [[] for _ in range(self.driver_num)]
        commit_results = []

        for r in self.requests:
            if r.selected_driver != -1:
                all_driver_req[r.selected_driver].append(r.RequestID)
        
        for l in range(self.driver_num):
            driver_json = json.dumps({"DriverID":l, "RequestList":all_driver_req[l], "LogicalClock":self.logical_clock})
            commit_results.append(driver_json)
        return commit_results

    def schedule(self, logical_clock: int, request_list: list, driver_statues: list) -> list:
        """
        add new keys ("now_sla", "type", "score") into request dict
        """
        for r in self.requests:
            r.now_sla -= 1
        
        requests_dict = [json.loads(i) for i in request_list]
        drivers_dict = [json.loads(i) for i in driver_statues]
        for r_d in requests_dict:
            r = ReqStructure()
            r.RequestID = r_d["RequestID"]
            r.RequestType = r_d["RequestType"]
            r.SLA = r_d["SLA"]
            r.Driver = r_d["Driver"]
            r.RequestSize = r_d["RequestSize"]
            r.LogicalClock = r_d["LogicalClock"]
            r.now_sla = r_d["SLA"]
            self.requests.append(r)

        assert len(drivers_dict) == self.driver_num
        self.driver_status = []
        for d_d in drivers_dict:
            d = DriverStructure()
            d.DriverID = d_d["DriverID"]
            d.Capacity = d_d["Capacity"]
            d.LogicalClock = d_d["LogicalClock"]
            self.driver_status.append(d)

        self.logical_clock = logical_clock
        
        for i in range(len(self.requests)-1, -1, -1):
            self.requests[i].type = self.set_type(self.requests[i])
            if self.requests[i].type == ABANDON:
                self.requests.remove(self.requests[i])
                continue
            self.requests[i].score = self.set_score(self.requests[i])
            self.requests[i].selected_driver = -1
        
        self.remain_cap = [d.Capacity for d in self.driver_status]
        self.real_cap = self.remain_cap
        # print(f'self.remain_cap: {self.remain_cap}') 

        self.sort()

        self.type_schedule(URGENT)
        self.type_schedule(noURGENT)
        
        excu = self.excu_reqs()

#         for r in self.requests:
#             print(f'"RequestID": {r.RequestID}, "RequestType": {r.RequestType}, "SLA": {r.SLA}, \
# "Driver": {r.Driver}, "RequestSize": {r.RequestSize}, "score": {r.score}, \
# "now_sla": {r.now_sla}, "type": {r.type}, "selected_driver": {r.selected_driver} ')

        for i in range(len(self.requests)-1, -1, -1):
            if self.requests[i].selected_driver != -1:
                self.requests.remove(self.requests[i])

        return excu
        